File: apps/hono-idp/src/queries/getUser.ts
import type { UsersRepository } from '../repositories/users';

export type PublicUser = {
  id: string;
  email: string;
  created_at: string;
  updated_at: string;
};

export function getUserQuery(users: UsersRepository) {
  return async (id: string): Promise<PublicUser | null> => {
    const user = await users.findById(id);
    if (!user) return null;
    return {
      id: user.id,
      email: user.email,
      created_at: user.createdAt.toISOString(),
      updated_at: user.updatedAt.toISOString(),
    };
  };
}

File: apps/go-idp/internal/queries/get_user.go
package queries

import (
	"context"

	"go-idp/internal/repositories"
)

type PublicUser struct {
	ID        string `json:"id"`
	Email     string `json:"email"`
	CreatedAt string `json:"created_at"`
	UpdatedAt string `json:"updated_at"`
}

type GetUser struct {
	users *repositories.UsersRepository
}

func NewGetUser(users *repositories.UsersRepository) *GetUser {
	return &GetUser{users: users}
}

func (q *GetUser) Handle(ctx context.Context, id string) (*PublicUser, error) {
	u, err := q.users.FindByID(ctx, id)
	if err != nil {
		return nil, err
	}
	if u == nil {
		return nil, nil
	}
	return &PublicUser{
		ID:        u.ID,
		Email:     u.Email,
		CreatedAt: u.CreatedAt.UTC().Format("2006-01-02T15:04:05.000Z"),
		UpdatedAt: u.UpdatedAt.UTC().Format("2006-01-02T15:04:05.000Z"),
	}, nil
}
